File: frontend/src/types/kryssord.ts
export interface KryssordStatistikk {
  antall_fullfort: number;
  antall_forsok: number;
}

export interface GridJson {
  celler: string[][];
  storrelse: number;
}

export interface LedetradOppslag {
  nr: number;
  tekst: string;
  ledetrad: string;
  kilde?: string;
  // Valgfrie posisjonsfelter – leveres ikke av API, avledet fra grid i frontend
  kol?: number;
  rad?: number;
  lengde?: number;
}

export interface LedetradJson {
  across: Record<string, LedetradOppslag>;
  down: Record<string, LedetradOppslag>;
}

export interface Kryssord {
  id: string;
  tittel: string;
  vanskelighetsgrad: "lett" | "middels" | "vanskelig";
  grid_storrelse: number;
  grid_json: GridJson;
  ledetrad_json: LedetradJson;
  opprettet_dato: string;
  publisert: boolean;
  statistikk: KryssordStatistikk | null;
}

export interface KryssordListeElement {
  id: string;
  tittel: string;
  vanskelighetsgrad: "lett" | "middels" | "vanskelig";
  grid_storrelse: number;
  opprettet_dato: string;
  antall_fullfort: number;
}

export interface KryssordArkiv {
  totalt: number;
  side: number;
  per_side: number;
  kryssord: KryssordListeElement[];
}

File: backend/app/schemas/kryssord.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class StatistikkUt(BaseModel):
    antall_fullfort: int
    antall_forsok:   int

    model_config = ConfigDict(from_attributes=True)


class KryssordUt(BaseModel):
    id:                UUID
    tittel:            str
    vanskelighetsgrad: str
    grid_storrelse:    int
    grid_json:         dict
    ledetrad_json:     dict
    opprettet_dato:    datetime
    publisert:         bool
    statistikk:        StatistikkUt | None = None

    model_config = ConfigDict(from_attributes=True)


class KryssordListeElement(BaseModel):
    id:                UUID
    tittel:            str
    vanskelighetsgrad: str
    grid_storrelse:    int
    opprettet_dato:    datetime
    antall_fullfort:   int = 0

    model_config = ConfigDict(from_attributes=True)


class KryssordArkiv(BaseModel):
    totalt:     int
    side:       int
    per_side:   int
    kryssord:   list[KryssordListeElement]


class GenererRequest(BaseModel):
    storrelse:  int = Field(default=13, ge=5, le=21)
    publiser:   bool = Field(default=False)
    tittel:     str | None = None


class GenererRespons(BaseModel):
    id:             UUID
    tittel:         str
    grid_storrelse: int
    antall_ord:     int
    publisert:      bool
